app/src/main/java/com/infomaniak/mail/utils/SharedValue.kt

```diff
@@ -0,0 +1,102 @@
+    }
+}
+
+fun SharedPreferences.sharedValue(key: String, defaultValue: Int): ReadWriteProperty<Any, Int> {
+    return object : ReadWriteProperty<Any, Int> {
+        override fun getValue(thisRef: Any, property: KProperty<*>): Int = getInt(key, defaultValue)
+        override fun setValue(thisRef: Any, property: KProperty<*>, value: Int) = transaction { putInt(key, value) }
+    }
+}
+
+fun SharedPreferences.sharedValue(key: String, defaultValue: Float): ReadWriteProperty<Any, Float> {
+    return object : ReadWriteProperty<Any, Float> {
+        override fun getValue(thisRef: Any, property: KProperty<*>): Float = getFloat(key, defaultValue)
+        override fun setValue(thisRef: Any, property: KProperty<*>, value: Float) = transaction { putFloat(key, value) }
+    }
+}
+
+fun SharedPreferences.sharedValue(key: String, defaultValue: Long): ReadWriteProperty<Any, Long> {
+    return object : ReadWriteProperty<Any, Long> {
+        override fun getValue(thisRef: Any, property: KProperty<*>): Long = getLong(key, defaultValue)
+        override fun setValue(thisRef: Any, property: KProperty<*>, value: Long) = transaction { putLong(key, value) }
+    }
+}
+
+fun SharedPreferences.sharedValue(key: String, defaultValue: String): ReadWriteProperty<Any, String> {
+    return object : ReadWriteProperty<Any, String> {
+        override fun getValue(thisRef: Any, property: KProperty<*>): String = getString(key, defaultValue) ?: defaultValue
+        override fun setValue(thisRef: Any, property: KProperty<*>, value: String) = transaction { putString(key, value) }
+    }
+}
+
+fun SharedPreferences.sharedValueNullable(key: String, defaultValue: String?): ReadWriteProperty<Any, String?> {
+    return object : ReadWriteProperty<Any, String?> {
+        override fun getValue(thisRef: Any, property: KProperty<*>): String? = getString(key, defaultValue)
+        override fun setValue(thisRef: Any, property: KProperty<*>, value: String?) = transaction { putString(key, value) }
+    }
+}
+
+fun SharedPreferences.sharedValue(key: String, defaultValue: Set<String>): ReadWriteProperty<Any, Set<String>> {
+    return object : ReadWriteProperty<Any, Set<String>> {
+        override fun getValue(thisRef: Any, property: KProperty<*>) = getStringSet(key, defaultValue) ?: defaultValue
+        override fun setValue(thisRef: Any, property: KProperty<*>, value: Set<String>) = transaction { putStringSet(key, value) }
+    }
+}
+
+fun SharedPreferences.sharedValue(key: String, defaultValue: List<String>): ReadWriteProperty<Any, List<String>> {
+    return object : ReadWriteProperty<Any, List<String>> {
+        override fun getValue(thisRef: Any, property: KProperty<*>): List<String> {
+            return getString(key, null)?.let(ApiController.json::decodeFromString) ?: defaultValue
+        }
+
+        override fun setValue(thisRef: Any, property: KProperty<*>, value: List<String>) {
+            transaction { putString(key, ApiController.json.encodeToString(value)) }
```

Comment:
I don't like that the json object we use here in in ApiController

Reply:
ApiController is just a static object holding properties, this `json` property is perfectly fine.